consul/server_manager/server_manager.go

```diff
@@ -0,0 +1,340 @@
+	serverConfigLock  sync.Mutex
+
+	// shutdownCh is a copy of the channel in consul.Client
+	shutdownCh chan struct{}
+
+	logger *log.Logger
+
+	// serf is used to estimate the approximate number of nodes in a
+	// cluster and limit the rate at which it rebalances server
+	// connections
+	clusterInfo ConsulClusterInfo
+
+	// notifyFailedServersBarrier is acts as a barrier to prevent
+	// queueing behind serverConfigLog and acts as a TryLock().
+	notifyFailedBarrier int32
+}
+
+// AddServer takes out an internal write lock and adds a new server.  If the
+// server is not known, appends the server to the list.  The new server will
+// begin seeing use after the rebalance timer fires or enough servers fail
+// organically.  If the server is already known, merge the new server
+// details.
+func (sm *ServerManager) AddServer(server *server_details.ServerDetails) {
+	sm.serverConfigLock.Lock()
+	defer sm.serverConfigLock.Unlock()
+	serverCfg := sm.getServerConfig()
+
+	// Check if this server is known
+	found := false
+	for idx, existing := range serverCfg.servers {
+		if existing.Name == server.Name {
+			newServers := make([]*server_details.ServerDetails, len(serverCfg.servers))
+			copy(newServers, serverCfg.servers)
+
+			// Overwrite the existing server details in order to
+			// possibly update metadata (e.g. server version)
+			newServers[idx] = server
+
+			serverCfg.servers = newServers
+			found = true
+			break
+		}
+	}
+
+	// Add to the list if not known
+	if !found {
+		newServers := make([]*server_details.ServerDetails, len(serverCfg.servers), len(serverCfg.servers)+1)
+		copy(newServers, serverCfg.servers)
+		newServers = append(newServers, server)
+		serverCfg.servers = newServers
+	}
+
+	sm.saveServerConfig(serverCfg)
+}
+
+// cycleServers returns a new list of servers that has dequeued the first
+// server and enqueued it at the end of the list.  cycleServers assumes the
+// caller is holding the serverConfigLock.
+func (sc *serverConfig) cycleServer() (servers []*server_details.ServerDetails) {
+	numServers := len(sc.servers)
+	if numServers < 2 {
+		return servers // No action required
+	}
+
+	newServers := make([]*server_details.ServerDetails, 0, numServers)
+	newServers = append(newServers, sc.servers[1:]...)
+	newServers = append(newServers, sc.servers[0])
+	return newServers
+}
+
+// FindServer takes out an internal "read lock" and searches through the list
+// of servers to find a "healthy" server.  If the server is actually
+// unhealthy, we rely on Serf to detect this and remove the node from the
+// server list.  If the server at the front of the list has failed or fails
+// during an RPC call, it is rotated to the end of the list.  If there are no
+// servers available, return nil.
+func (sm *ServerManager) FindServer() *server_details.ServerDetails {
+	serverCfg := sm.getServerConfig()
+	numServers := len(serverCfg.servers)
+	if numServers == 0 {
+		sm.logger.Printf("[WARN] consul: No servers found in the server config")
```

Comment:
Same here - I'd probably drop this since it's at a pretty low level and could be spammy.